print-macsec.c

```diff
@@ -0,0 +1,202 @@
+
+#define MACSEC_SECTAG_LEN_NOSCI 6
+#define MACSEC_SECTAG_LEN_SCI 14
+static int
+ieee8021ae_sectag_len(netdissect_options *ndo, const struct macsec_sectag *sectag)
+{
+	return (GET_U_1(sectag->tci_an) & MACSEC_TCI_SC) ?
+	       MACSEC_SECTAG_LEN_SCI :
+	       MACSEC_SECTAG_LEN_NOSCI;
+}
+
+static int macsec_check_length(netdissect_options *ndo, cconst struct macsec_sectag *sectag, u_int length, u_int caplen)
+{
+	u_int len;
+
+	/* we need the full MACsec header in the capture */
+	if (caplen < (MACSEC_SECTAG_LEN_NOSCI + 2))
+		return 0;
+
+	len = ieee8021ae_sectag_len(ndo, sectag);
+	if (caplen < (len + 2))
+		return 0;
+
+	if ((GET_U_1(sectag->short_length) & MACSEC_SL_MASK) != 0) {
+		/* original packet must have exact length */
+		u_int exact = ETHER_HDRLEN + len + 2 + (GET_U_1(sectag->short_length) & MACSEC_SL_MASK);
+		return exact == length;
+	} else {
+		/* original packet must not be short */
+		u_int minlen = ETHER_HDRLEN + len + 2 + 48;
+		return length >= minlen;
+	}
+
+	return 1;
+}
+
+#define SCI_FMT "%016" PRIx64
+
+static const struct tok macsec_flag_values[] = {
+	{ MACSEC_TCI_E,   "E" },
+	{ MACSEC_TCI_C,   "C" },
+	{ MACSEC_TCI_ES,  "S" },
+	{ MACSEC_TCI_SCB, "B" },
+	{ MACSEC_TCI_SC,  "I" },
+	{ 0, NULL }
+};
+
+/* returns < 0 iff the packet can be decoded completely */
+int macsec_print(netdissect_options *ndo, const u_char **bp,
+		 u_int *lengthp, u_int *caplenp, u_int *hdrlenp,
+		 u_short *length_type)
+{
+	const u_char *p = *bp;
+	u_int length = *lengthp;
+	u_int caplen = *caplenp;
+	u_int hdrlen = *hdrlenp;
+	const struct macsec_sectag *sectag = (const struct macsec_sectag *)p;
+	u_int len;
+
+	if (!macsec_check_length(sectag, length, caplen)) {
+		ND_PRINT((ndo, tstr));
+		return hdrlen + caplen;
+	}
+
+	if (sectag->unused || sectag->tci_an & MACSEC_TCI_VERSION) {
+		ND_PRINT((ndo, "%s", istr));
+		return hdrlen + caplen;
+	}
+
+	if (ndo->ndo_eflag) {
+		char buf[128];
+		int n = snprintf(buf, sizeof(buf), "an %u, pn %u, flags %s",
+				 GET_U_1(sectag->tci_an) & MACSEC_AN_MASK,
+				 GET_BE_U_4(sectag->packet_number),
+				 bittok2str_nosep(macsec_flag_values, "none",
+						  GET_U_1(sectag->tci_an) & MACSEC_TCI_FLAGS));
+		if (n < 0)
+			return hdrlen + caplen;
+
+
+		if ((GET_U_1(sectag->short_length) & MACSEC_SL_MASK) != 0 {
+			int r = snprintf(buf + n, sizeof(buf) - n, ", sl %u",
+					 GET_U_1(sectag->short_length) & MACSEC_SL_MASK);
+			if (r < 0)
+				return hdrlen + caplen;
+			n += r;
+		}
+
+		if (GET_U_1(sectag->tci_an) & MACSEC_TCI_SC) {
+			uint64_t sci;
+			int r;
+			sci = GET_BE_U_8(sectag->secure_channel_id);
+			r = snprintf(buf + n, sizeof(buf) - n, ", sci " SCI_FMT, sci);
+			if (r < 0)
+				return hdrlen + caplen;
+			n += r;
+		}
+
+		ND_PRINT((ndo, "%s, ", buf));
+	}
+
+	len = ieee8021ae_sectag_len(ndo, sectag);
+	*length_type = GET_BE_U_2(*bp + len);
+	if (ndo->ndo_eflag && *length_type > ETHERMTU && !(GET_U_1(sectag->tci_an) & MACSEC_TCI_E))
```

Comment:
print-ether.c uses `MAX_ETHERNET_LENGTH_VAL`, defined in tcpdump's <ethertype.h>, instead of `ETHERMTU` which might be in the system's <net/ethernet.h>

Reply:
I'll read the spec to see why an 802.1 standard would *need* to know about a particular 802.x network's maximum packet size.

Reply:
The `length_type` can either be a frame length, if it's equal to or smaller than the MTU, or it's an ethertype.  I don't quite know why it's relevant in this branch, but it's checking that length_type contains an ethertype.

Reply:
> The `length_type` can either be a frame length, if it's equal to or smaller than the MTU, or it's an ethertype. I don't quite know why it's relevant in this branch, but it's checking that length_type contains an ethertype.

It's no longer doing so; it's now punting that to the Ethernet dissector, which is as it should be.

Reply:
Excellent!